indico/modules/categories/controllers/display.py

```diff
@@ -685,24 +685,63 @@ def _render_ongoing_events(self, ongoing_events):
         return template.render_ongoing_events(ongoing_events, self.category.display_tzinfo)
 
 
-class RHCategoryUpcomingEvent(RHDisplayCategoryBase):
-    """Redirect to the upcoming event of a category."""
+class RHCategoryRelativeEventBase(RHDisplayCategoryBase):
+    """Redirect to a close event in a category."""
+
+    _no_event_message = None
 
     def _process(self):
-        event = self._get_upcoming_event()
+        event = self._get_event()
         if event:
             return redirect(event.url)
         else:
-            flash(_('There are no upcoming events for this category'))
+            flash(self._no_event_message)
             return redirect(self.category.url)
 
-    def _get_upcoming_event(self):
+    def _get_query_opts(self):
+        raise NotImplementedError
+
+    def _get_event(self):
+        filter_, order = self._get_query_opts()
         query = (Event.query
                  .filter(Event.is_visible_in(self.category.id),
-                         Event.start_dt > now_utc(),
+                         filter_,
                          ~Event.is_deleted)
                  .options(subqueryload('acl_entries'))
-                 .order_by(Event.start_dt, Event.id))
+                 .order_by(*order))
         res = get_n_matching(query, 1, lambda event: event.can_access(session.user))
         if res:
             return res[0]
+
+
+class RHCategoryPreviousEvent(RHCategoryRelativeEventBase):
+    """Redirect to the previoue event of a category."""
+
+    _no_event_message = _('There are no upcoming events for this category')
```

Comment:
Copy-paste?

Reply:
oops, yes. fixed and also fixed a typo in the docstring ;)